packages/storage/src/lib.rs

```diff
@@ -431,58 +484,99 @@ where
     }
 }
 
-impl<S: StorageBacking, T: Serialize + DeserializeOwned + Clone + Send + Sync> Deref
-    for StorageEntry<S, T>
-{
+impl<P: StoragePersistence<Option<T>>, T: Send + Sync> Deref for StorageEntry<P, T> {
     type Target = Signal<T>;
 
     fn deref(&self) -> &Signal<T> {
         &self.data
     }
 }
 
-impl<S: StorageBacking, T: Serialize + DeserializeOwned + Clone + Send + Sync> DerefMut
-    for StorageEntry<S, T>
-{
+impl<P: StoragePersistence<Option<T>>, T: Send + Sync> DerefMut for StorageEntry<P, T> {
     fn deref_mut(&mut self) -> &mut Signal<T> {
         &mut self.data
     }
 }
 
-impl<S: StorageBacking, T: Display + Serialize + DeserializeOwned + Clone + Send + Sync> Display
-    for StorageEntry<S, T>
-{
+impl<P: StoragePersistence<Option<T>>, T: Display> Display for StorageEntry<P, T> {
     fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
         self.data.fmt(f)
     }
 }
 
-impl<S: StorageBacking, T: Debug + Serialize + DeserializeOwned + Clone + Send + Sync> Debug
-    for StorageEntry<S, T>
-{
+impl<P: StoragePersistence<Option<T>>, T: Debug> Debug for StorageEntry<P, T> {
     fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
         self.data.fmt(f)
     }
 }
 
 /// A trait for a storage backing
-pub trait StorageBacking: Clone + 'static {
+pub trait StorageBacking<T>: 'static {
+    type Encoder: StorageEncoder<T>;
+    type Persistence: StoragePersistence<
+            Option<T>,
+            Value = Option<<Self::Encoder as StorageEncoder<T>>::EncodedValue>,
+        >;
+
+    /// Gets a value from storage for the given key
+    fn get(key: &<Self::Persistence as StoragePersistence<Option<T>>>::Key) -> Option<T> {
+        let loaded = Self::Persistence::load(key);
+        match loaded {
+            // TODO: this treats None the same as failed decodes.
+            Some(x) => {
+                let deserialized = Self::Encoder::deserialize(&x);
+                if let Err(err) = &deserialized {
+                    warn!("Deserialization error: {err:?}");
+                }
+                deserialized.ok()
+            }
+            None => {
+                warn!("Got None for key {key:?}");
```

Comment:
You can actually hit this if you delete the local storage entry while the web page is open. That will go on to crash on an unwrap since the current API surface doesn't handle it.

I don't think this is a regression: the old code use to unwrap before it got to this point.

In the future, we should probably have some policy of using the initial value in both empty and error cases, but have an option for the user to provide a custom handler or something. That seems like an unrelated feature and I'm considering it out of scope for now. My changes do mage it easier to implement by plumbing the errors and missing cases up to this layer however.